keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -883,20 +889,51 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
-which no node exists in the cluster) or scheduled (for example when they fail
-to pull the the docker image). In case of non-scheduled pods they will be moved
-to the `Failed` phase by the PodGC controller as it marks as failed all
-non-scheduled terminating pods (after the Alpha changes for this feature, see
-above). We plan to extend the PodGC controller to also move to the failed phase
-(and delete) scheduled pods in the pending state which are terminating.
+A scheduled pod (i.e. a pod is bound to a node and its `spec.nodeName` indicates
+the node) may also get stuck in a `Pending` phase (for example due to an invalid
+image reference). In that case, as long as the pod is not terminating, it
+could be a transient issue and Kubelet can make progress after retrying.
+Once such a Pod (in a `Pending` phase) is also terminating, Kubelet is not going
+to attempt starting of the pod. Here, we say that a pod is terminating when its
+`deletionTimestamp` is set by API server due to a Delete request (which can be
+called on the pod by a user or a k8s component).
+
+Such cases are problematic in case of Jobs with pod failure policies, which we want
+to match only against pods in terminal phase (`Succeeded` or `Failed`), to ensure
+their state is not modified while Job controller matches them against the policy.
+
+In order to ensure that Job controller only considers pods in terminal phase we
+need to make sure pods in the following states are transitioned into `Failed`
+phase:
+1. `Running` phase, and orphaned (scheduled to a node which no longer exists)
+2. `Pending` phase, terminating, and unscheduled
+3. `Pending` phase, terminating, and scheduled
+
+For Alpha, we implemented a fix for scenarios 1. and 2. by setting the pod phase
+as `Failed` in PodGC. Note that, in both scenarios there is no Kubelet responsible
+to transition the pod, thus it is done by PodGC.
+
+As the 3rd scenario wasn't fixed in the first iteration of Beta (1.26), we only
+require pods to be in terminal phase when `podFailurePolicy` is specified (and
+only in that case), but this creates an inconsistency with handling Jobs
+with and without pod failure policies
+(see Issue: [Job controller should wait for Pods to terminate to match the failure policy](https://github.com/kubernetes/kubernetes/issues/113855)).
+
+For the second iteration of Beta (1.27), we plan to fix the 3rd scenario by
+modifying Kubelet to transition `Pending`, terminating, and scheduled pods into
+the `Failed` phase, allowing the Job
+controller to count them, match against the (optional) pod failure policy and
+remove their Job finalizers, allowing API server to complete deletion.
+A prototype implementation is prepared here:
+[Mark Pending Terminating pods as Failed by Kubelet](https://github.com/kubernetes/kubernetes/pull/115331).
+
+Also, for the second iteration of Beta (1.27) we are going to expand the
+feature documentation to explain this change.
```

Comment:
It may be worth to note here some of the cases that pod may be stuck in pending. In general for a pod to go into pending all the containers have to been started. So if a container does not start, for example the init containers did not succeeded or one of the containers could not start, it will remain in pending. xref: https://github.com/kubernetes/kubernetes/blob/d9f21e55df4cd95f5e6c9ea1d7109a2a767920fb/pkg/kubelet/kubelet_pods.go#L1408-L1418

Reply:
Thanks, updated the point about documentation, to provide the list of example scenarios.